File: 0443-string-compression/0443-string-compression.java
class Solution {
    public int compress(char[] chars) {
        int index = 0; // where to write
        int i = 0;     // where to read

        while (i < chars.length) {
            char current = chars[i];
            int count = 0;

            // count how many times current character repeats
            while (i < chars.length && chars[i] == current) {
                i++;
                count++;
            }

            // write the character
            chars[index++] = current;

            // if count > 1, write each digit of count
            if (count > 1) {
                for (char c : Integer.toString(count).toCharArray()) {
                    chars[index++] = c;
                }
            }
        }

        return index; // length of compressed array
    }
}

File: 0443-string-compression/0443-string-compression.cpp
class Solution {
public:
    int compress(vector<char>& chars) {
        vector<char> ans;
        int n = chars.size();
        int index = 0;
        for (int i = 0; i < chars.size(); i++) {
            char ch = chars[i];
            int cnt = 1;

            while (i < chars.size() - 1 && chars[i] == chars[i + 1]) {
                cnt++;
                i++;
            }
            chars[index++] = ch;
            if (cnt > 1) {
                string cntst = to_string(cnt);
                for (char c : cntst) {
                    chars[index++] = c;
                }
            }
        }
        return index;
    }
};